testplan/testing/multitest/suite.py

```diff
@@ -112,30 +112,55 @@ class may end up having different tag indices if they
 def get_testsuite_name(suite):
     """
     Returns the name to be used for the given testsuite. This is made of
-    either the class name or the result of `custom_name` (can be a normal
-    string or a function returning a string) if it exists. The first time
-    this function is called the suite name will be saved for future use.
+    either the class name or the result of `name` (can be a normal string
+    or a function returning a string) if it exists. The first time this
+    function is called the suite name will be saved for future use.
 
     :param suite: Suite object whose name is needed
     :type suite: ``testsuite``
 
     :return: Name of given suite
     :rtype: ``str``
     """
-    if hasattr(suite, "__name__"):
-        return suite.__name__
-
-    if suite.__class__.custom_name is None:
-        suite.__name__ = suite.__class__.__name__
-    elif isinstance(suite.__class__.custom_name, six.string_types):
-        suite.__name__ = suite.__class__.custom_name
-    elif callable(suite.__class__.custom_name):
-        suite.__name__ = suite.custom_name(suite.__class__.__name__)
+    if "name" in suite.__dict__:
+        return suite.name
+
+    if suite.__class__.name is None:
+        suite.name = suite.__class__.__name__
+    elif isinstance(suite.__class__.name, six.string_types):
+        suite.name = suite.__class__.name
+    elif callable(suite.__class__.name):
+        # In Python2 unbound method cannot be called by class directly
+        suite.name = six.get_method_function(suite.name)(
+            suite.__class__.__name__, suite
+        )
     else:
-        # Should not reach here, something must be wrong
-        assert False, "Unknown `custom_name` of [{}]".format(suite.__class__)
+        raise RuntimeError("Should not reach here: in {}".format(suite))
+
+    if not isinstance(suite.name, six.string_types):
+        raise ValueError(
+            'Test suite name "{name}" must be a string, it is of type:'
+            " {type}".format(name=suite.name, type=type(suite.name))
+        )
+    elif not suite.name:
+        raise ValueError("Test suite name cannot be an empty string")
 
-    return suite.__name__
+    suite.name = six.ensure_str(suite.name)
+
+    if len(suite.name) > defaults.MAX_TESTSUITE_NAME_LENGTH:
```

Comment:
originally I have such a function but removed, for some reasons:
1. The warning messages are not same, e.g. Multitest object, Testsuite object, testcase.
2. No check for colon in testcase, because last time we just change it back to slice the pattern to at most 3 parts, so testcase name can have colon.
3. length of the names are not necessary the same, although we make them same now.